src/renderer/atlas/shader_ps.hlsl

```diff
@@ -67,6 +68,87 @@ Output main(PSData data) : SV_Target
         color = weights * data.color;
         break;
     }
+    case SHADING_TYPE_TEXT_BUILTIN_GLYPH:
+    {
+        // The RGB components of builtin glyphs are used to control the generation of pixel patterns in this shader.
+        // Below you can see their intended effects where # indicates lit pixels.
+        //
+        // .r = stretch
+        //      0: #_#_#_#_
+        //         _#_#_#_#
+        //         #_#_#_#_
+        //         _#_#_#_#
+        //
+        //      1: #___#___
+        //         __#___#_
+        //         #___#___
+        //         __#___#_
+        //
+        // .g = invert
+        //      0: #_#_#_#_
+        //         _#_#_#_#
+        //         #_#_#_#_
+        //         _#_#_#_#
+        //
+        //      1: _#_#_#_#
+        //         #_#_#_#_
+        //         _#_#_#_#
+        //         #_#_#_#_
+        //
+        // .r = fill
+        //      0: #_#_#_#_
+        //         _#_#_#_#
+        //         #_#_#_#_
+        //         _#_#_#_#
+        //
+        //      1: ########
+        //         ########
+        //         ########
+        //         ########
+        //
+        float4 glyph = glyphAtlas[data.texcoord];
+        float2 pos = floor(data.position.xy / (thinLineWidth * data.renditionScale));
+
+        // A series of on/off/on/off/on/off pixels can be generated with:
+        //   step(frac(x * 0.5f), 0)
+        // The inner frac(x * 0.5f) will generate a series of
+        //   0, 0.5, 0, 0.5, 0, 0.5
+        // and the step() will transform that to
+        //   1,   0, 1,   0, 1,   0
+        //
+        // We can now turn that into a checkerboard pattern quite easily,
+        // if we imagine the fields of the checkerboard like this:
+        //   +---+---+---+
+        //   | 0 | 1 | 2 |
+        //   +---+---+---+
+        //   | 1 | 2 | 3 |
+        //   +---+---+---+
+        //   | 2 | 3 | 4 |
+        //   +---+---+---+
+        //
+        // Because this means we just need to set
+        //   x = pos.x + pos.y
+        // and so we end up with
+        //   step(frac(dot(pos, 0.5f)), 0)
+        //
+        // Finally, we need to implement the "stretch" explained above, which can
+        // be easily achieved by simply replacing the factor 0.5 with 0.25 like so
+        //   step(frac(x * 0.25f), 0)
+        // as this gets us
+        //   0, 0.25, 0.5, 0.75, 0, 0.25, 0.5, 0.75
+        // = 1,    0,   0,    0, 1,    0,   0,    0
+        //
+        // Of course we only want to apply that to the X axis, which means
+        // below we end up having 2 different multipliers for the dot().
+        float stretched = step(frac(dot(pos, float2(glyph.r * -0.25f + 0.5f, 0.5f))), 0) * glyph.a;
+        // Thankfully the remaining 2 operations are a lot simpler.
+        float inverted = abs(glyph.g - stretched);
+        float filled = max(glyph.b, inverted);
+
+        color = premultiplyColor(data.color) * filled;

```

Comment:
nit:

Can we do this instead (if this is legal in hlsl)?

```hlsl
int on = step(frac(dot(pos, float2(glyph.r * -0.25f + 0.5f, 0.5f))), 0) * glyph.a;
on ^= glyph.g;
on |= glyph.b;

color = premultiplyColor(data.color) * on;
```

Reply:
It is legal in hlsl, but this won't work here. `glyph.b` may not just be 0 or 1, because some of the box drawing and powerline glyphs (╳, etc.) are anti-aliased and have semi-transparent pixels.
Also, the `abs()` and `max()` calls are super efficient on GPUs.

Side note: The most expensive part by far in the shader (like probably 10x over the above 4 lines) is the switch/case because GPUs don't seem to support computed jumps (or jump tables). That means that reaching the 7th case label needs at a minimum 7 jumps which are not fast. On AMD cards it needs twice that (you can test that with https://shader-playground.timjones.io/ and by setting the compiler to the Radeon GPU Analyzer).
We have the option to stop using an "uber shader" like this and instead use multiple distinct pixel shaders that each do one (or few) of the `ShadingType`s.
A middle ground could be to manually split up the switch/case into multiple smaller switches. For instance, our pixel shader can clearly be split into 3 parts: "depends on the background texture", "depends on the glyphAtlas texture" and "depends on nothing". I've experimented with that and it lead to a significant size reduction of the shader, but I've not tested how much power we can save that way. (On my desktop GPU I found that it's impossible to measure that, because it's just way too fast for our pixel shader either way.)

Reply:
Isn't this ░ ▒ ▓ specific code (especially, the `step(frac(dot(...)))` part? If we're not going to work with floats (either in future or now), I would recommend to just simplify it to use int instead of float as it makes the code more readable 😅

*(we can also get rid of the extra `glyph.a` multiplication in the `stretched` calc as it is always going to be 1)*

Reply:
Initially, I used the new `ShadingType` only for the 3 shaded glyphs ░ ▒ ▓ and I used the `ShadingType::TextGrayscale` for the other box drawing glyphs. This allowed me to use integer operations, just like you suggested it. But I just didn't like this approach in the end for two reasons:
* 4 of the "Symbols for Legacy Computing" glyphs require both, shaded textures and anti-aliasing. They're U+1FB9C to U+1FB9F. For instance: https://www.compart.com/en/unicode/U+1FB9C
* The floating point logic generates more compact HLSL assembly and often runs faster on older/slower hardware too.

So, now the new shading type is used for all builtin glyphs and uses floating point, as this solves the 1st point. It's also why I need to multiply by `glyph.a`. It turns the 0/1 values that the `step()` call spits out into a 0/alpha value. I'm using a multiplication there, because this results in better assembly.

> I would recommend to just simplify it to use int instead of float as it makes the code more readable 😅

In particular the 2nd point is important to me, even more so than for our CPU-side code. On the currently upcoming 8K@60Hz displays this shader is supposed to churn out 2B pixels per second. The float math I wrote compiles down to more efficient instructions and saves us a couple cycles. It's not much but I think every bit helps. (...which is also why I'll probably have to investigate my "side note" above at some point or another.)